src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -557,6 +557,13 @@ internal StackSemaphore(int initialCount, JoinableTaskContext joinableTaskContex
                             }
                         }
 
+                        if (this.faulted)
+                        {
+                            // The semaphore faulted while we were waiting on it.
+                            DisposeReleaserNoThrow(releaser);
+                            this.ThrowIfFaulted();
+                        }
```

Comment:
Can't this all just be the one statement: `this.ThrowIfFaulted();` ?
Doing that will throw if faulted, and we're in a try whose finally guarantees to call `DisposeReleaserNoThrow` already, and in fact won't do anything other than that (since pushed ==false) at this point.
For that matter, is this actually required for the fix? It seems like the crux of the fix may just be that you wrapped the `if (pushed)` block in its own `try` block.

Reply:
True , the `try` block may be all that is needed. The `this.ThrowIfFaulted();` will reduce the number of `operation()` executions after the semaphore is faulted though. How important do you think that is?

Reply:
I suspect running `operation()` when the semaphore is already faulted is a bad idea. So I'm in favor of keeping the `ThrowIfFaulted()` code here. I just think the rest of it should be removed.

Reply:
I see you've already removed it. :)